### DISCUSSION TITLE
Error Handling Utility

### DISCUSSION THREAD

**Discussion by ryanlong1004** [ufs-community/uwtools#59] [Review and Planning RAP]
## Scope

The error handling utility will be used by every other UW Tool in the tool kit. It will standardize the way errors are handled and reported, and will allow users to treat errors and warnings at the level of severity required by the task at hand. 

This utility will be the primary method of handling errors within the UW Framework. It will be a utility used by all other workflow tools, and will be closed to modification, but open to extension.

Our number 1 goal with this feature is to ensure that it works toward a standardized output.

## Design
Error messages should provide, at minimum, the following output:

1. Time stamp
2. Caller
3. Data structure(s) involved
4. Environment info

## Questions
- Can we use a 3rd party library for this?
- Where does the tool live?

**Comment by aerorahul** [comment]
`logging.exception` is a standard way of "logging" with an exception handling that I have seen/used in the past.  There might be better other ways.
`logging.exception` will put a `time-stamp` and a `Traceback` so that should likely satisfy requirements 1 and 2.
I am not sure what 3. means. I think I understand what 4 is, but a dump of for e.g. `printenv` is a bit too much to have.

**Comment by christinaholtNOAA** [comment]
The uwtools package now uses standard logging, defined by the `logging` module, and handles uw-specific exceptions defined in the `exceptions` module.
